_posts/2025-02-13-rfprs.md

```diff
@@ -0,0 +1,372 @@
+4. Commit descriptions should discuss that commit's changes as necessary.
+
+Here's an example of a medium-sized PR that follows these guidelines:
+
+    Set ECS Attribute to indicate health of ledger latency
+
+    JIRA: F8N-365
+
+    # Overview
+
+    This is the first step of a longer term goal to be able to control the behavior of ECS task
+    placement for tasks which rely on this database. The idea is that if the task requires a healthy
+    ledger latency we can avoid placing it on a node which is behind.
+
+    # Approach
+
+    This PR adds some new behavior to the reflectors. The reflectors will start a ledger latency
+    monitor when they start. This monitor will periodically look at the ledger latency as reported
+    by the db client. If below a max threshold, it will ensure that the container instance on which
+    it runs has an ECS attribute to indicate the ledger has fallen behind.
+
+    # Next Steps
+
+    After this is deployed and working we should spot check new instances to ensure that they are
+    not marked as 'healthy' if they are still working on updating the ledger. Once we are confident
+    in this, we can then update our deployment tooling to lean on this new attribute.
+
+The commits on this PR try to follow RFPR guidelines. Notice how low value and high value changes
+are isolated:
+
+    * Move loop helper code to pkg/loop.
+    * Add ECS dependencies.
+    * Update executive interface mock.
+    * Create an ECSClient interface and generate mock.
+    * Create EcsMetada type to unmarshal ECS metadata.
+    * Create ledger monitor to set ECS 'unhealthy' attribute when falling behind.
+    * Add functional options for the Monitor.
+    * Update reflector config to also compose ledger health config.
+    * Add tests to verify attribute is set correctly using a mocked ECS environment.
+
+Example of what I consider to be a reasonable commit within this PR.
+
+    Create ledger monitor to set ECS 'unhealthy' attribute when falling behind.
+
+    The monitor follows the pattern that the other types use, where a Start(ctx) method blocks, and
+    only exits when the context is canceled.
+
+    It follows this high level logic:
+
+    * Every time the *time.Ticker ticks:
+        * get the ledger latency from the db
+        * compare this value to threshold
+        * if latency is OK, set a 'healthy' attr
+        * else set 'unhealthy' attr
+
+    It keepds the last known successful ECS attr set in a local var so that it doesn't set the attr
+    on each tick. The variable is a *bool because we need three states:
+
+    1. unknown state of ECS attr
+    2. 'healthy'
+    3. 'unhealthy'
+
+
+These are the kinds of PRs I'm thankful to get notified about. I might spend on average 15-30m extra
+to create PRs using these guidelines, but it saves many hours across the reviewers assigned to it,
+and the quality of feedback you get on your RFPR will invariably be higher.
```

Comment:
Any PRs out there on public repos that could be linked to as examples?

Reply:
asked in slack